neps/nep-0491.md

```diff
@@ -0,0 +1,386 @@
+  "Transfer": { "deposit": "100" },
+  "AddKey": { "public_key": "...", "access_key": "..." }
+}
+```
+
+To create implicit accounts, the current protocol requires a single `Transfer`
+action without further actions in the same transaction and this has not changed
+with this proposal:
+
+```json
+"Actions": {
+  "CreateAccount": {},
+  "ReserveStorage": { "deposit": "1000000000000000000000000" },
+  "Transfer": { },
+}
+```
+
+If a non-refundable transfer arrives at an account that already exists, it will
+fail and the funds are returned to the predecessor.
+
+Finally, when querying an account for its balance, there will be an additional
+field `nonrefundable` in the output. Wallets will need to decide how they want
+to show it. They could, for example, add a new field called "non-refundable
+storage credits".
+
+```js
+// Account near
+{
+  "amount": "68844924385676812880674962949",
+  "block_hash": "3d6SisRc5SuwrkJnLwQb3W5pWitZKCjGhiKZuc6tPpao",
+  "block_height": 97314513,
+  "code_hash": "Dmi6UTRYTT3eNirp8ndgDNh8kYk2T9SZ6PJZDUXB1VR3",
+  "locked": "0",
+  "storage_paid_at": 0,
+  "storage_usage": 2511772,
+  "formattedAmount": "68,844.924385676812880674962949",
+  // this is new
+  "nonrefundable": "0"
+}
+```
+
+
+## Reference Implementation
+
+On the protocol side, we need to add new action:
+
+```rust
+enum Action {
+  CreateAccount(CreateAccountAction), 
+  DeployContract(DeployContractAction),
+  FunctionCall(FunctionCallAction),
+  // kept for backwards compatibility but new clients should only use `TransferV2`
+  Transfer(TransferAction),
+  Stake(StakeAction),
+  AddKey(AddKeyAction),
+  DeleteKey(DeleteKeyAction),
+  DeleteAccount(DeleteAccountAction),
+  Delegate(super::delegate_action::SignedDelegateAction),
+  // this gets added in the end
+  ReserveStorage(ReserveStorageAction),
+}
+```
+
+and handle the new action in the `apply_action` call.
+
+Further, we have to update the account meta data representation in the state
+trie to track the non-refundable storage.
+
+```rust
+pub struct Account {
+    amount: Balance,
+    locked: Balance,
+    // this field is new
+    nonrefundable: Balance,
+    code_hash: CryptoHash,
+    storage_usage: StorageUsage,
+    // the account version will be increased from 1 to 2
+    version: AccountVersion,
+}
+```
+
+The field `nonrefundable` must be added to the normal `amount` and the `locked`
+balance calculate how much state the account is allowed to use. The new formula
+to check storage balance therefore becomes
+
+```rust
+amount + locked + nonrefundable >= storage_usage * storage_amount_per_byte
+```
+
+For old accounts that don't have the new field, the non-refundable balance is
+always zero. Adding non-refundable balance later is not allowed. If a transfer
+is made to an account that already existed before the receipt's actions are
+applied, execution must fail with
+`ActionErrorKind::OnlyImplicitAccountCreationAllowed{ account_id: AccountId }`.
```

Comment:
Should this be a new error kind instead? Something like `ActionErrorKind::OnlyReserveStorageOnAccountCreation`

Reply:
Ah you are right, it should be a new error since it can be triggered independently of `ActionErrorKind::OnlyImplicitAccountCreationAllowed`.